cudapoa/src/cudapoa_nw_tb_banded.cuh

```diff
@@ -0,0 +1,586 @@
+                {
+                    last_score = first_element_prev_score;
+                }
+
+                if (score.s0 < last_score + gap_score)
+                {
+                    score.s0 = last_score + gap_score;
+                    trace.t0 = 0;
+                }
+
+                if (score.s1 < score.s0 + gap_score)
+                {
+                    score.s1 = score.s0 + gap_score;
+                    trace.t1 = 0;
+                }
+
+                if (score.s2 < score.s1 + gap_score)
+                {
+                    score.s2 = score.s1 + gap_score;
+                    trace.t2 = 0;
+                }
+
+                int32_t tscore = max(score.s2 + gap_score, score.s3);
+                if (tscore > score.s3)
+                {
+                    score.s3 = tscore;
+                    trace.t3 = 0;
+                    loop     = true;
+                }
+            }
+
+            // Copy over the last element score of the last lane into a register of first lane
+            // which can be used to compute the first cell of the next warp.
+            first_element_prev_score = __shfl_sync(FULL_MASK, score.s3, WARP_SIZE - 1);
+
+            int64_t index      = static_cast<int64_t>(read_pos + 1 - band_start) + static_cast<int64_t>(score_gIdx % score_matrix_height) * static_cast<int64_t>(band_width + CUDAPOA_BANDED_MATRIX_RIGHT_PADDING);
+            scores[index]      = score.s0;
+            scores[index + 1L] = score.s1;
+            scores[index + 2L] = score.s2;
+            scores[index + 3L] = score.s3;
+
+            index                 = static_cast<int64_t>(read_pos + 1 - band_start) + static_cast<int64_t>(score_gIdx) * static_cast<int64_t>(band_width + CUDAPOA_BANDED_MATRIX_RIGHT_PADDING);
+            traceback[index]      = trace.t0;
+            traceback[index + 1L] = trace.t1;
+            traceback[index + 2L] = trace.t2;
+            traceback[index + 3L] = trace.t3;
+
+            __syncwarp();
+        }
+    }
+
+    int32_t aligned_nodes = 0;
+    if (lane_idx == 0)
+    {
+        // Find location of the maximum score in the matrix.
+        int32_t i      = 0;
+        int32_t j      = read_length;
+        int32_t mscore = min_score_value;
+
+        for (int32_t idx = 1; idx <= graph_count; idx++)
+        {
+            if (outgoing_edge_count[graph[idx - 1]] == 0)
+            {
+                if ((graph_count - idx) < score_matrix_height)
+                {
+                    int32_t s = get_score_tb(scores, idx, j, score_matrix_height, gradient, band_width, max_column, min_score_value);
+                    if (mscore < s)
+                    {
+                        mscore = s;
+                        i      = idx;
+                    }
+                }
+                else
+                {
+                    ///ToDo throw an error indicating selected score_matrix_height (i.e. max predecessor distance) is too small
```

Comment:
is this still necessary?

Reply:
not sure, can we have a case where `end node (with 0 outgoing edges) < (graph_count - some_positive_int)`? if so, choosing `score_matrix_height <= some_positive_int` will be problematic.

Reply:
it could be - I'm picturing a case where the graph has several end nodes, each with outgoing edges = 0. only one of them is part of the backtrace which has the highest score. in that case the matrix height >= positive int. so let's return an error if this fails

Reply:
I thought a bit and found it's cleaner if I add a new error code and address this in the other PR which in includes both adaptive and static traceback NW.

Reply:
Okay, let's do that.